### PR TITLE
Handle projection of parametric BlockOperators

### LINKED ISSUE #784: Projecting parametric block operators
Supersedes #778.

### PR DESCRIPTION
**Pull request by sdrave**
Fixes #784.

### CODE REVIEW

**Comment by renefritze** [comment]
https://github.com/pymor/pymor/pull/785#issuecomment-543040263

File: src/pymor/algorithms/projection.py [hunk 1]
```diff
@@ -198,10 +197,32 @@ def action_LincombOperator(self, op):
...
 
-    @match_class(OperatorInterface)
-    def action_generic_projection(self, op):
-        op.logger.warning('Using inefficient generic projection operator')
-        return ProjectedOperator(op, self.range_basis, self.source_basis, self.product)
+    @match_class(BlockOperatorBase)
+    def action_BlockOperatorBase(self, op):
+        if op.blocked_range:
+            if self.range_basis is not None:
+                range_basis = self.range_basis._blocks
```
Comment (pmli) [inline src/pymor/algorithms/projection.py]: How about this?
```suggestion
                range_bases = self.range_basis._blocks
```

File: src/pymor/algorithms/projection.py [hunk 2]
```diff
@@ -198,10 +197,32 @@ def action_LincombOperator(self, op):
...
+
+        projected_ops = np.array([[project(op.blocks[i, j], rb, sb)
+                                   for j, sb in enumerate(source_basis)]
+                                  for i, rb in enumerate(range_basis)])
+        if self.range_basis is None and op.blocked_range:
+            return BlockColumnOperator(np.sum(projected_ops, axis=1))
+        elif self.source_basis is None and op.blocked_source:
+            return BlockRowOperator(np.sum(projected_ops, axis=0))
+        else:
+            return np.sum(projected_ops)
```
Comment (pmli) [inline src/pymor/algorithms/projection.py]: Is it guaranteed that `self.range_basis` and `self.source_basis` are not both `None`?

File: src/pymor/algorithms/projection.py (hunk 1 above)
Comment (sdrave) [reply src/pymor/algorithms/projection.py]: Changed.

File: src/pymor/algorithms/projection.py (hunk 2 above)
Comment (sdrave) [reply src/pymor/algorithms/projection.py]: See #791.

**Pull request merged**